### PR TITLE
Implement property page support for Project Query API

### PR DESCRIPTION
**Pull request by tmeschter**
Implement the property page sections of the Project Query API defined in CPS. There's a lot of code here, but a relatively small number of patterns that I will explain shortly. First, some background.

The Project Query API defines the following types that relate to property pages and their values:

```
Project
  -> PropertyPage[]
    -> Category[]
    -> UIProperty[]
      -> UIPropertyValue[]
        -> ConfigurationDimension[]
        -> SupportedValue[]
      -> UIPropertyEditor[]
        -> UIEditorMetadata[]
```

Read this as "a `Project` has multiple `PropertyPages`, each `PropertyPage` has multiple `UIProperties`, each `UIProperty` has multiple `UIPropertyValues` (one per configuration)" and so on. Each of these types has additional properties (sometimes many) that constitute the actual data you want to retrieve using the Project Query API; for brevity they are not shown here.

The Project Query API supports two ways of retrieving one of these model types: by ID, and by relationship. In the case of ID we may be asked to produce a `UIProperty` for the property with ID "A" on the page with ID "B" that is part of project "C:\D\E\F.csproj". On the other hand, when asked to retrieve model types by relationship, we might be given a `PropertyPage` that has already been created and asked to produce all of the related `UIProperties`. CPS is responsible for examining a query and deciding which parts of the requested data to retrieve by ID, and which to retrieve by relationship. Our responsibility is to implement the actual retrieval mechanisms for each model type.

The types added here fall into one of four patterns. The `FooProvider` types exist at the boundary between the CPS parts of the implementation, and our parts. Each provider specifies, through metadata, which parts of the model it support, and which of the two retrieval mechanisms it supports. All the providers here support "by relationship" retrieval, but not all of them support by ID retrieval because not all of the model types have a meaningful ID. They delegate to other types to do the actual work, and we describe those types next.

At the bottom we have a series of static `FooProducer` types. These are responsible for taking our internal data structures--or IDs which map to internal data structures--and producing the related "Foo" model. If the model has a meaningful ID, it is also responsible for creating the hierarchical identity--usually by starting with the parent's identity and tagging on another item. Another thing to note about these types is that the Project Query API supports retrieving only some of the properties on any given model. The ones that should be retrieved are specified by a `FooPropertiesAvailableStatus` object, and a lot of what this producer does is simply check the status of each relevant property, and copy data into the model if it is requested.

The `FooByIdProducer` types implement the ID retrieval. Given a hierarchical identity, they extract the relevant pieces and pass them on to the methods on the related `FooProducer` type to produce the "FooModel". The `FooByIdProducer` type then reports the new models back to the lower layers of the Project Query API.

The `FooFromBarProducer` types implement the "by relationship" retrieval. Given some other model object "Bar", they examine the `ProviderState` field (used to pass data from one producer to another during a query execution) for the internal data representing "Bar", pass it on to `FooProducer` to handle, and again report the results back to the Project Query API.

The other interesting part of this change is the `PropertyPageQueryCache`, which serves to cache data that may be accessed multiple times in the course of the query. Please see the doc comment on that type for more details. To simplify unit tests, this is always accessed through the related `IPropertyPageQueryCache` interface.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/project-system/pull/6615)

### CODE REVIEW

**Review by drewnoakes — COMMENTED** [review]
Gave it a look through and left a few comments. Will review again more thoroughly once it's ready for merge.

As always, thanks for the thorough description. It helped to see beyond the ceremony required by the query system 👍

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Query/PropertyPages/CategoryDataProvider.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    /// <remarks>
+    /// Responsible for populating <see cref="IPropertyPage.Categories"/>. Can also retrieve a <see cref="ICategory"/>
+    /// based on its ID.
+    /// </remarks>
+    [QueryDataProvider(CategoryType.TypeName, ProjectModel.ModelName)]
+    [RelationshipQueryDataProvider(PropertyPageType.TypeName, PropertyPageType.CategoriesPropertyName)]
+    [QueryDataProviderZone(ProjectModelZones.Cps)]
+    [Export(typeof(IQueryByIdDataProvider))]
+    [Export(typeof(IQueryByRelationshipDataProvider))]
+    internal partial class CategoryDataProvider : QueryDataProviderBase, IQueryByIdDataProvider, IQueryByRelationshipDataProvider
```
[the remark this answers is not in the document]
Comment (tmeschter) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Query/PropertyPages/CategoryDataProvider.cs:27]: Yeah, that's a left over from when the producers were private nested classes within the providers. I'll remove that.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Query/PropertyPages/CategoryByIdDataProducer.cs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+                if (requestId.KeysCount == 3
+                    && requestId.TryGetValue(ProjectModelIdentityKeys.ProjectPath, out string path)
+                    && requestId.TryGetValue(ProjectModelIdentityKeys.PropertyPageName, out string propertyPageName)
+                    && requestId.TryGetValue(ProjectModelIdentityKeys.CategoryName, out string categoryName))
+                {
+                    try
+                    {
+                        if (_projectService.GetLoadedProject(path) is UnconfiguredProject project
+                            && await project.GetProjectLevelPropertyPagesCatalogAsync() is IPropertyPagesCatalog projectCatalog
+                            && projectCatalog.GetSchema(propertyPageName) is Rule rule)
```
[the remark this answers is not in the document]
Comment (tmeschter) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Query/PropertyPages/CategoryByIdDataProducer.cs:45]: I didn't realize I could do that. Yes, that's a better representation of my intent.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Query/PropertyPages/PropertyPageQueryCache.cs [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+            _catalogCache = new Dictionary<ProjectConfiguration, IPropertyPagesCatalog?>();
+            _ruleCache = new Dictionary<(ProjectConfiguration, string), IRule?>();
+        }
+
+        /// <summary>
+        /// Retrieves the set of <see cref="ProjectConfiguration"/>s for the project.
+        /// </summary>
+        public Task<IImmutableSet<ProjectConfiguration>?> GetKnownConfigurationsAsync() => _knownProjectConfigurations.GetValueAsync();
+
+        private async Task<IImmutableSet<ProjectConfiguration>?> CreateKnownConfigurationsAsync()
```
[the remark this answers is not in the document]
Comment (tmeschter) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Query/PropertyPages/PropertyPageQueryCache.cs:62]: Ah, I thought I'd gotten all of those.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Query/PropertyPages/CategoryByIdDataProducer.cs [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        {
+            Requires.NotNull(projectService, nameof(projectService));
+            _projectService = projectService;
+        }
+
+        public async Task SendRequestAsync(QueryProcessRequest<IReadOnlyCollection<EntityIdentity>> request)
+        {
+            Requires.NotNull(request, nameof(request));
+
+            foreach (var requestId in request.RequestData)
```
[the remark this answers is not in the document]
Comment (tmeschter) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Query/PropertyPages/CategoryByIdDataProducer.cs:34]: I'm not seeing any warnings for those...

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Query/PropertyPages/CategoryByIdDataProducer.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (tmeschter) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Query/PropertyPages/CategoryByIdDataProducer.cs:45]: That syntax produces errors for me.

Note that I'm doing two things here: defining the `project`, `projectCatalog`, and `rule` locals, and checking that the values are not null. I don't seem to be able to combine the `not null` syntax with the local declaration.

[the remark this answers is not in the document]
Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Query/PropertyPages/CategoryByIdDataProducer.cs:45]: I though this had been completed but it looks like it's still on the road map for C# 9: https://github.com/dotnet/csharplang/issues/1350

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Query/PropertyPages/CategoryByIdDataProducer.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Query/PropertyPages/CategoryByIdDataProducer.cs:34]: I would expect IDE0008 based on:

https://github.com/dotnet/project-system/blob/e79cda6d4985ab489123b99b94cca24888819047/.editorconfig#L136-L139

Just checked and these appear as messages in the error list, not warnings.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Query/PropertyPages/CategoryByIdDataProducer.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (tmeschter) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Query/PropertyPages/CategoryByIdDataProducer.cs:45]: I'm going to leave this as-is for now.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Query/PropertyPages/UIPropertyValueDataProducer.cs [hunk 5]
```diff
@@ -1,4 +1,4 @@
-﻿// Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE.md file in the project root for more information.
+﻿ // Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE.md file in the project root for more information.
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Query/PropertyPages/UIPropertyValueDataProducer.cs]: nit whitespace

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Utilities/EnumerableExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+        /// <typeparam name="T"></typeparam>
+        /// <param name="enumerable"></param>
+        /// <returns></returns>
+        public static IEnumerable<(int index, T item)> WithIndices<T>(this IEnumerable<T> enumerable)
+        {
+            int index = 0;
+            foreach (var item in enumerable)
+            {
+                yield return (index++, item);
+            }
```
[the remark this answers is not in the document]
Comment (tmeschter) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Utilities/EnumerableExtensions.cs:22]: Yes, it should be possible, but I'm not going to worry about it right now.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Query/PropertyPages/PropertyPageQueryCache.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (tmeschter) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Query/PropertyPages/PropertyPageQueryCache.cs:34]: I ran into issues with nullable when I tried doing this--the compiler didn't like that I was trying to convert `Task<IImmutableSet<ProjectConfiguration>>` to `Task<IImmutableSet<ProjectConfiguration?>>`. I'm going to leave as-is for the moment.

**Review by jjmew — APPROVED** [review]
:shipit:

**Pull request merged**